### PR TITLE
[C/C++] improved syntax, constants are detected in global scope

### PR DESCRIPTION
**Pull request by dpronin**
C/C++ syntax is improved, constants are detected in global scope

Signed-off-by: Denis Pronin <dannftk@yandex.ru>

### CODE REVIEW

**Comment by deathaxe** [comment]
Would it be valid for ObjC/ObjC++ as well?

Those are intented to be kept inline with C/C++ whenever possible.

**Comment by dannftk** [comment]
@deathaxe, nothing has been touched or spoiled specifically to ObjC and ObjC++. As you could notice the syntax only touched and extended, these changes should work for ObjC++ as well

**Comment by dannftk** [comment]
Could someone proceed progress of the PR? I have no idea what I should do to continue merging process?

**Comment by deathaxe** [comment]
What I wanted to say is: Changes to C/C++ package should also be applied to ObjC/ObjC++ if possible.

**Comment by dannftk** [comment]
@deathaxe , you may be convinced, they are

**Comment by deathaxe** [comment]
Sorry, I am not.

In short: Adding the test case of this PR after https://github.com/sublimehq/Packages/blob/43081b3494a3d743b9aeb4b60e971b3f1f5528eb/Objective-C/syntax_test_objc%2B%2B.mm#L2759 causes those to fail as `kMyConstant` is not scoped `constant`.

The reason is the `global-type` context of _C++/C++.sublime-syntax_ modified by this PR not to be shared with _Objective-C/ObjC++.sublime-syntax_ automatically.

In order to keep ObjC++ in sync with C++, modifications of this PR need to be copied to ObjC++ as well.

_Maybe ObjC/ObjC++ will be implemented as inherited syntax of C/C++ to enable automatic feature sharing at some point in future, but it is a long path to go to achieve that._

**Comment by dannftk** [comment]
@deathaxe, check it out, I've spread the changes on ObjC++

**Comment by deathaxe** [comment]
Thanks!

**Comment by dannftk** [comment]
@deathaxe , who is going to merge these changes?

**Comment by deathaxe** [comment]
A member of sublimehq, I guess. @BenjaminSchaaf announced to maintain syntax review. But with all those pending core stability tasks pending to fix remaining early-after-release bugs I guess it might take a while.

**Comment by dannftk** [comment]
How to request someone to run CI on the PR?

File: C++/C++.sublime-syntax [hunk 1]
```diff
@@ -945,7 +945,7 @@ contexts:
       scope: keyword.operator.c++
     - match: '(?=\b({{control_keywords}}|{{operator_keywords}}|{{casts}}|{{memory_operators}}|{{other_keywords}}|operator)\b)'
       pop: true
-    - match: '(?=\s)'
+    - match: '(?=\s+)'
```
Comment (deathaxe) [inline C++/C++.sublime-syntax:948]: Why adding `+`? It's a lookahead, which wouldn't consume it anyway.

File: Objective-C/Objective-C++.sublime-syntax [hunk 2]
```diff
@@ -1058,7 +1058,7 @@ contexts:
     - match: '(?={{path_lookahead}}\s*::\s*$)'
       set: [function-definition-params, global-function-identifier]
     - match: '(?=\S)'
-      fail: global-maybe
+      fail: true
```
Comment (deathaxe) [inline Objective-C/Objective-C++.sublime-syntax:1061]: Do you want to `pop`?

File: C++/C++.sublime-syntax (hunk 1 above)
Comment (dannftk) [reply C++/C++.sublime-syntax:948]: Indeed, I missed that it is look ahead and does not make any sense. What do you think if we ate here all the whitespace symbols by a normal regexp?

File: Objective-C/Objective-C++.sublime-syntax (hunk 2 above)
Comment (dannftk) [reply Objective-C/Objective-C++.sublime-syntax:1061]: Yes, otherwise it leads to infinite cycling in case both branches global-maybe fail

Comment (dannftk) [reply Objective-C/Objective-C++.sublime-syntax:1061]: I will fix to pop in 2-3 hours

File: C++/C.sublime-syntax [hunk 3]
```diff
@@ -838,25 +841,28 @@ contexts:
               pop: true
             - include: preprocessor-disabled
 
+  preprocessor-if-true:
```
Comment (deathaxe) [inline C++/C.sublime-syntax:844]: While createing named contexts is generally a good thing, it appears unrelated with the topic if this PR.

File: C++/C.sublime-syntax [hunk 4]
```diff
@@ -838,25 +841,28 @@ contexts:
...
-          pop: true
-        - match: ^\s*(#\s*else)\b
-          captures:
-            1: meta.preprocessor.c keyword.control.import.else.c
-          push:
-            - meta_content_scope: comment.block.preprocessor.else-branch.c
-            - match: (?=^\s*#\s*endif\b)
-              pop: true
-            - include: preprocessor-disabled
+        - include: preprocessor-if-true
```
Comment (deathaxe) [inline C++/C.sublime-syntax:865]: In such situations a direct `push: preprocessor-if-true` would be preferable, because it would also apply any `meta_include_prototype`, `meta_scope` or `meta_content_scope` directive from the pushed context.

File: C++/C++.sublime-syntax [hunk 5]
```diff
@@ -275,19 +275,19 @@ contexts:
       push: global-modifier
     # Take care of comments just before a function definition.
     - match: /\*
-      scope: punctuation.definition.comment.c
+      scope: punctuation.definition.comment.c++
```
Comment (deathaxe) [inline C++/C++.sublime-syntax:278]: I'd vote to keep PRs small and focused on certain aspects. It is probably not always easy but still preferable.

This PR is about detecting "global constants".

**Review by deathaxe — CHANGES_REQUESTED** [review]

**Comment by dannftk** [comment]
Is there anything that prevents the PR from approving?

**Comment by deathaxe** [comment]
Mostly patience, I guess. Most of us are volenteers spending our sparetime.

**Review by deathaxe — APPROVED** [review]

**Comment by dpronin** [comment]
ping

**Comment by deathaxe** [comment]
Could you please resolve the conflicts?

**Comment by dpronin** [comment]
@deathaxe done

**Review by deathaxe — APPROVED** [review]

**Review by jrappen — APPROVED** [review]

File: C++/C++.sublime-syntax [hunk 6]
```diff
@@ -1129,6 +1129,23 @@ contexts:
     - match: (?=\W)
       pop: true
 
+  global-maybe-constant-or-function:
+    - include: comments
+    - include: types
+    - include: modifiers
+    - include: constants
+    - include: variables
+    - include: scope:source.c#named-constants
```
Comment (deathaxe) [inline C++/C++.sublime-syntax:1138]: The included `constants` context of line 1136 already includes `scope:source.c#comments` which itself includes `named-constants`. So this include seems duplicated and should be removed.

Doing so would also make this context obsolete as the pattern could be moved back into _constants_.

File: Objective-C/Objective-C++.sublime-syntax [hunk 7]
```diff
@@ -1014,6 +1014,23 @@ contexts:
     - match: (?=\W)
       pop: true
 
+  global-maybe-constant-or-function:
+    - include: comments
+    - include: types
+    - include: modifiers
+    - include: constants
+    - include: variables
+    - include: scope:source.c#named-constants
```
Comment (deathaxe) [inline Objective-C/Objective-C++.sublime-syntax:1023]: Same here.

File: C++/C.sublime-syntax [hunk 8]
```diff
@@ -163,14 +163,17 @@ contexts:
...
   constants:
     - match: \b(__func__|NULL|true|false|TRUE|FALSE)\b
       scope: constant.language.c
     - match: \b(__FILE__|__FUNCTION__|__LINE__)\b
       scope: support.constant.c
-    # common C constant naming idiom -- kConstantVariable
-    - match: '\bk[A-Z]\w*\b'
-      scope: constant.other.variable.mac-classic.c
+    - include: named-constants
     - match: \b(noErr|kNilOptions|kInvalidID|kVariableLengthArray)\b
```
Comment (deathaxe) [inline C++/C.sublime-syntax:177]: Could be reverted with the other two references of `named-constants` removed.

**Comment by dpronin** [comment]
May I ask what have changed with code basically since your last approval for you to leave 3 comments to resolve?

**Comment by dpronin** [comment]
The PR has been on for 8 months and none got interested

**Comment by deathaxe** [comment]
Not the code, but just noticed those details after another look. The detail about comment scopes was noticed due to work on the linked PR. The other ones I found by luck. 

I do the best I can to try to judge quality of contributions with the level of insight I have in a syntax. I don't use most of them actively, so I haven't C/C++ for about 15 years. Maintaining these packages is just some kind of spare time soduko alternative.

C/C++ family is one of the not actively maintained syntaxes, atm. A PR needs 2 approvals to be able to be merged. So it isn't enough for one to care about - maybe for good reasons to limit biases.

_IMHO, one reason for missing interest might be an earlier announcements of planes to rewrite the syntax. The announcing dev has moved on to other projects. Noone else has been keen enough to start this very large undertaking to rewrite 4 very complex syntaxes, while others might still just wait for it to happen judging all intermediate PRs not worth it._
